File: frontend/src/lib/preflight.ts
// Pure helpers for the Queue tab's Preflight box (docs/transfers-redesign-spec.md §4,
// prefigured; this task's own handoff prompt, prompts/done/2026-08-20-preflight-box.md, plus its
// follow-up prompts/2026-08-20-preflight-waiting-sources.md) -- "something lftpweb already knows
// about but has no work to do on yet." Kept source-agnostic, matching `PreflightRowOut` itself
// (`api/types.ts`): the *arr poller and the settle gate's own eligibility check are the two
// sources wired up (a settle row reads "remote — 22 GB", per `preflightSizeLabel` below) --
// nothing here assumes a row always came from the *arr. This project's whole component-testing
// story is `lib/*.test.ts` (README.md's Known gaps: no component rendering is tested), so the
// box's own display logic lives here, unit-tested, rather than inlined in
// `components/PreflightBox.tsx`.

import type { PreflightRowOut, SettleSettingsOut } from '../api/types'
import { formatBytes, formatEta, settleWaitLabel } from './format'

/** The page-size selector's own default -- "5 rows by default" is the user's own original
 * words for this box, unchanged by the selector's arrival (2026-08-21, "we should have a drop
 * down on preflight like the rest, show 5/10/20 etc"): a box whose whole point is a handful of
 * not-yet-arrived releases still starts small, it just no longer needs a separate "Show all"
 * expand to see more -- the selector replaces that control rather than sitting beside it.
 */
export const PREFLIGHT_DEFAULT_PAGE_SIZE = 5

/** The selector's own offered sizes -- **5/10/20, not the other two boxes' 10/20/50**
 * (`lib/pagination.ts.PAGE_SIZE_OPTIONS`): this box is smaller by intent (a handful of
 * not-yet-arrived releases, not a growing job history), so its own "see more at once" ceiling
 * is smaller too. A dedicated list and validator rather than a re-export of the shared one,
 * since the two boxes' option sets are allowed to diverge -- `isPreflightPageSize` is this
 * box's own `lib/pagination.ts.isPageSize`.
 */
export const PREFLIGHT_PAGE_SIZE_OPTIONS = [5, 10, 20] as const

export type PreflightPageSize = (typeof PREFLIGHT_PAGE_SIZE_OPTIONS)[number]

/** Validates a page size read back out of `localStorage` -- `lib/pagination.ts.isPageSize`'s
 * own contract, mirrored for this box's own narrower option list: a hand-edited value, a
 * foreign one, or a size this version no longer offers must fall back to the default rather
 * than being trusted.
 */
export function isPreflightPageSize(value: unknown): value is PreflightPageSize {
  return typeof value === 'number' && (PREFLIGHT_PAGE_SIZE_OPTIONS as readonly number[]).includes(value)
}

/** A row's own size figure, when its source provided one. "`NN% of X`" once both a total and a
 * remaining figure are known (an *arr row still downloading); just the total when only that is
 * known (the settle-gate follow-up's own expected shape -- a release already fully present
 * remotely has nothing "left", just a known size); `null` when the source gave neither --
 * **never a placeholder for a source that didn't report one** (the handoff prompt's own "never a
 * request to enrich it").
 */
export function preflightSizeLabel(
  row: Pick<PreflightRowOut, 'size_bytes' | 'size_remaining_bytes'>,
): string | null {
  const total = row.size_bytes
  const remaining = row.size_remaining_bytes
  if (total == null || total <= 0) return null
  if (remaining == null) return formatBytes(total)
  const done = Math.max(0, total - remaining)
  const percent = Math.round((done / total) * 100)
  return `${percent}% of ${formatBytes(total)}`
}

/** Capitalizes a source's own free-form status text for display (`"downloading"` ->
 * `"Downloading"`) -- shown verbatim otherwise, never interpreted or mapped to a fixed
 * vocabulary (unlike `chipStateFor`'s own job-state chip): a second source's own wording
 * (`"Settling"`, say) must render exactly as that source wrote it. `null` straight through, so a
 * row with nothing to say renders nothing rather than a placeholder dash.
 */
export function preflightStatusLabel(statusLabel: string | null): string | null {
  if (!statusLabel) return null
  return statusLabel.charAt(0).toUpperCase() + statusLabel.slice(1)
}

/** The figure column's own text (2026-08-21, "we missed the remaining time") -- `preflightSizeLabel`
 * above, plus a " · <duration> left" suffix once `remaining_s` is known, through the exact
 * `formatEta` + " left" combination `lib/transferPanel.ts.transferLineValue` already uses for the
 * Transfers row's own ETA, rather than a second time-formatting idiom. A settle-gated row's
 * `remaining_s` is always `null` (`core/preflight.py.PreflightRow`'s own docstring: that source's
 * remaining figure is `size_bytes`, not a time) so it renders exactly as `preflightSizeLabel`
 * alone always has -- this function widens what an *arr row can show without changing a settle
 * row's own reading at all.
 */
export function preflightRemainingLabel(
  row: Pick<PreflightRowOut, 'size_bytes' | 'size_remaining_bytes' | 'remaining_s'>,
): string | null {
  const base = preflightSizeLabel(row)
  if (row.remaining_s == null) return base
  const remaining = `${formatEta(row.remaining_s)} left`
  return base ? `${base} · ${remaining}` : remaining
}

/** An *arr row's own `trackedDownloadState`, translated into what it means **to lftpweb** rather
 * than shown as the *arr's raw word (2026-08-21, user's own browser report: "that downloading
 * chip in preflight is arr telling it is downloading from the client ... I think we should spell
 * that out"). `downloading` -- the state the user actually saw -- becomes **"Waiting"**
 * (2026-08-21, a follow-up browser look: an earlier "Waiting for download" round read equally as
 * "lftpweb is waiting to download it" and "waiting for the download client to finish"; the user
 * considered "Waiting for remote client", asked to shorten it, then chose this from a set of
 * options). Deliberately short and silent on *where*:
 *
 * - **7 characters**, well under the 20 "Waiting for download" cost, so it can never crowd the
 *   row (this box's own figure column is already the widest on the page, `w-44`).
 * - Matches its sibling chip, `Settling`, in shape exactly -- one word, present tense, a state.
 *   The two are the only chips this box ever shows, and now read as one vocabulary.
 * - Says nothing about *where* the waiting is happening, and doesn't need to: the *arr brand
 *   logo sits in the same row, the box itself is called "Preflight", and the tooltip
 *   (`preflightChipTooltip` below) already spells out `Downloading from "<client>" -- reported
 *   by <instance>`. Rejected alternatives, worth keeping on record: "At client" (says where, but
 *   mild jargon), "Downloading" (the *arr's own word -- precisely the wording that caused the
 *   original confusion), "Grabbed" (reads as a past event, sitting oddly beside a present-tense
 *   `Settling`).
 *
 * `importing` (`core/arrclient.py.TRACKED_DOWNLOAD_STATE_IMPORTING`, the one other state that
 * codebase's own module docstring records as verified against a live Sonarr) keeps its own word,
 * **"Importing"** -- a release in this state has already finished arriving and is a meaningfully
 * different situation from "still waiting," so collapsing it into the same word would be
 * actively wrong, not just imprecise. Anything else falls through to `preflightStatusLabel`'s
 * existing "render the source's own wording verbatim" rule -- **deliberately not translated
 * further**, per the handoff prompt's own "do not invent a state the *arr does not report": this
 * codebase has only verified those two `trackedDownloadState` values live, so a third
 * (paused/stalled/queued at the download client, say) gets shown honestly rather than guessed at.
 */
const ARR_CHIP_LABELS: Record<string, string> = {
  downloading: 'Waiting',
  importing: 'Importing',
}

export function preflightChipLabel(row: Pick<PreflightRowOut, 'source' | 'status_label'>): string | null {
  if (row.source !== 'arr') return preflightStatusLabel(row.status_label)
  if (!row.status_label) return null
  return ARR_CHIP_LABELS[row.status_label] ?? preflightStatusLabel(row.status_label)
}

/** The chip's own hover text (2026-08-21). An *arr row's own detail and its provenance
 * (user's own words: "tooltip maybe we should show the arr details. Downloading from
 * '<download client name>' from arr") -- kept off the chip's visible text (which now speaks
 * lftpweb's own vocabulary, `preflightChipLabel` above) and onto hover instead.
 * `download_client` is read straight from the *arr's own response (`core/arrsync.py`,
 * `QueueRecord.raw["downloadClient"]`); `source_label` is already this row's own *arr instance
 * name. Falls back to naming just the instance when the *arr didn't report a download client,
 * rather than an empty quote. `null` only for a status-less *arr row -- there is nothing to
 * attribute yet.
 *
 * **A non-*arr (settle) row gets a tooltip too** (2026-08-21, follow-up: "the settling chip
 * should have a mouse over that shows time details") -- the box was asymmetric before this,
 * since only the *arr chip had one, and "Settling" alone says nothing about how much longer.
 * Rather than inventing a second countdown sentence, this reuses `lib/format.ts.
 * settleWaitLabel` **verbatim** -- the exact "Waiting for changes -- 1 of 2 scans, 35s of 60s"
 * wording the Files tree's own state text and the lifecycle R-icon tooltip already share, so
 * all three agree rather than a third copy of that sentence drifting from the other two. Fed
 * `row.wait_scans`/`wait_since` (`core/preflight.py.PreflightRow`'s own generic fields, `null`
 * for an *arr row) as the *counts*, not a pre-baked string, so the countdown recomputes live
 * from `Date.now()` on every render rather than freezing at whatever the last poll returned --
 * `settleWaitLabel` itself owns that arithmetic. `settle` is the site-wide `GET
 * /api/settings/settle` constants (`required_scans`/`min_age_s`), `null` until that one fetch
 * resolves -- `settleWaitLabel` degrades to the bare "Waiting for changes" either way, never
 * blocking the tooltip from rendering. Always non-null for a non-*arr row (`settleWaitLabel`
 * itself never returns `null`), unlike the *arr branch above.
 */
export function preflightChipTooltip(
  row: Pick<
    PreflightRowOut,
    'source' | 'status_label' | 'download_client' | 'source_label' | 'wait_scans' | 'wait_since'
  >,
  settle: SettleSettingsOut | null,
): string | null {
  if (row.source !== 'arr') {
    return settleWaitLabel(
      { settle_matched_scans: row.wait_scans, settle_first_matched_at: row.wait_since },
      settle,
    )
  }
  if (!row.status_label) return null
  return row.download_client
    ? `Downloading from "${row.download_client}" — reported by ${row.source_label}`
    : `Reported by ${row.source_label}`
}

File: backend/lftpweb/core/preflight.py
"""The Preflight box's source-agnostic shape (docs/transfers-redesign-spec.md §4, prefigured;
this task's own handoff prompt, prompts/done/2026-08-20-preflight-box.md) -- "something lftpweb
already knows about but has no work to do on yet," independent of *which* upstream told it so.

**Two sources are wired up: the *arr poller (`core/arrsync.py`), and the settle gate's own
eligibility check (`core/autoqueue.py.AutoQueue`, added by
prompts/2026-08-20-preflight-waiting-sources.md).** A settle-gated row differs from an *arr row
in an important way: it *does* have a remote presence and a known remote size (it reads
"remote — 22 GB"), whereas an *arr queue record has no remote presence at all yet. **Nothing in
this module may name *arr, the settle gate, or any other single source, by construction** --
`PreflightRow.source` is the one place a caller learns which source a row came from, and
`source_label`/`source_kind`/`status_label` are free-form, source-owned display text this module
never interprets. Keeping that boundary here (rather than baking either source's own vocabulary
into the shared row/cache shape) is what let the settle-gate source add itself without reshaping
anything the first task shipped -- see `docs/decisions.md` for how that held up in practice.

`PreflightHold` is the flap-tolerance cache a source uses when its own report can go briefly
missing for reasons unrelated to the underlying fact changing -- `core/arrsync.py` is the one
user of it: a row missing from a poll for up to `PREFLIGHT_HOLD_S` keeps showing rather than
blinking out and back in, because a download client's own queue can blank out for a beat (the
*arr's own SABnzbd production incident, `core/arrsync.py`'s module docstring, 2026-08-18).
**Retirement is the one thing that skips the hold entirely** (2026-08-21, "a handed-over release
lingers in Preflight for up to 150s"): `update`'s own `retired` set is for a row a source knows is
gone for a *known* reason -- handed over to a real item -- as opposed to merely missing from this
one pass for a reason the source can't tell apart from a blip. See `update`'s own docstring for
the two-bucket split; the caller (`core/arrsync.py._preflight_candidates`) is where that
distinction is actually drawn, never here. **Not every source needs this.** `core/autoqueue.py`'s settle-gated rows are computed fresh from this
same process's own persisted state on every successful scan pass, with no external flakiness to
smooth over, so that source replaces its rows wholesale each pass instead of holding them here --
see that module's own "Preflight" section for why a full replace is strictly more correct for a
source with no flap risk to guard against. A row not refreshed within the hold window is deleted
from the cache outright, never merely marked stale -- there is no persisted state and no further
escalation, so this can never itself become a second accumulation risk on top of whatever a given
source already guards against on its own.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass
from typing import Literal

# Widened, not replaced, when a new source lands -- see this module's own docstring.
PreflightSource = Literal["arr", "settle"]


@dataclass(frozen=True)
class PreflightRow:
    """One Preflight box row -- "something we know about but have no work to do on yet."
    Deliberately thin and source-agnostic: no id, no bytes-done, no queue position -- there is no
    `item` and no `job` behind a row here, on any source, and the handoff prompt's own "the rows
    are inert, and the box is what makes that structural" is exactly why nothing here invites a
    per-row control that would need one.
    """

    source: PreflightSource
    queue_id: int
    # The bound queue's own display identity (2026-08-21, "the columns moved around" fix) --
    # common to every source, so it lives here rather than being re-derived per source or, worse,
    # left off entirely (the defect this task fixes: a Preflight row had `queue_id` but no name,
    # so it couldn't show the queue tag every other row on the page shows). `queue_name` is the
    # queue's full `name`; `queue_short_name` mirrors `api/settings_queues.py.
    # resolve_queue_display_name`'s own `short_name` field (`None` for "no short name set") so
    # the frontend's existing `lib/queueDisplayName.ts.queueDisplayName(short, name)` fallback
    # renders this row's tag identically to every Transfers row's own tag, with no second
    # fallback rule invented here.
    queue_name: str
    queue_short_name: str | None
    title: str
    # Free-form, source-owned display text for "what state is this in" -- an *arr row's own
    # `trackedDownloadState` (e.g. `"downloading"`), a settle-gate row's own wording (e.g.
    # `"Settling"`). Never interpreted or branched on by this module.
    status_label: str | None
    # The upstream's own display name (an *arr instance's configured name, e.g. `"Sonarr"`) and,
    # when the source has one, a brand/variant hint for the row's own chip (`'sonarr'`/`'radarr'`
    # for *arr; `None` for a source with no logo of its own -- the settle gate, say). Rendering
    # owns the `source`/`source_kind` -> logo mapping; this module only ever carries the strings.
    source_label: str
    source_kind: str | None
    # A known total size and, when the source can compute one, how much is left to arrive --
    # both `None` when the source has neither (never a request to enrich one that lacks it, per
    # the handoff prompt's own instruction). An *arr row (still downloading) can populate both
    # from the *arr's own `size`/`sizeleft`; a settle-gate row (already fully present remotely,
    # just being confirmed stable) is expected to populate only `size_bytes`, from `item.
    # remote_size` -- there is nothing "left" from that source's own point of view.
    size_bytes: int | None
    size_remaining_bytes: int | None
    # How many seconds until this row's own source expects its wait to clear -- generic in name
    # (unlike `size_bytes`/`size_remaining_bytes` above, it happens to be populated by exactly
    # one source today, the same way those two happen to be populated by both): an *arr row
    # reads its own queue record's `timeleft` (`core/arrsync.py._parse_timeleft`); a settle-gated
    # row leaves this `None` -- the gate's remaining wait is bound by *scan count*, not a wall-
    # clock estimate this codebase has any business fabricating (`core/autoqueue.py`'s own
    # "Preflight" section), and its already-existing remaining figure is `size_bytes` above, not
    # a time. `None` for "the source has no meaningful estimate this pass" -- never a fabricated
    # or zero figure (the handoff prompt's own instruction), so a caller renders nothing rather
    # than a wrong number.
    remaining_s: float | None
    # The download client actually fetching this release, from the *arr's own point of view
    # (2026-08-21, user's own words: "tooltip maybe we should show the arr details. Downloading
    # from '<download client name>' from arr") -- an *arr row's own `downloadClient`; `None` for
    # a settle row (there is no separate download client in that source's own model -- lftpweb
    # *is* the thing fetching it) and for an *arr row whose response didn't happen to carry one.
    # Display-only provenance for a chip tooltip, never branched on.
    download_client: str | None
    # A generic "how far along has this row's own wait gotten" detail for the chip's own hover
    # tooltip (2026-08-21, "the settling chip should have a mouseover that shows time details") --
    # deliberately named after neither condition a caller might compute it from, the same way
    # `size_bytes`/`remaining_s` above are shared vocabulary rather than either source's own
    # words. A row whose wait is bound by *scan count* populates both: `wait_scans` is how many
    # consecutive matching observations have been made so far, `wait_since` is the ISO-8601
    # timestamp the current streak began at (`core/autoqueue.py.on_scan` populates both from
    # `core/settle.py`'s own `matched_scans`/`updated_at` pair). Rendered client-side through
    # `lib/format.ts.settleWaitLabel` -- the exact "Waiting for changes -- 1 of 2 scans, 35s of
    # 60s" wording the Files tree and the lifecycle R-icon tooltip already share, given the site-
    # wide scan-count/wall-clock constants (`GET /api/settings/settle`) that helper also needs --
    # so a third copy of that sentence is never written here or anywhere else. An *arr row's own
    # wait isn't bound by scan count at all (`remaining_s` above already covers what it can say),
    # so it leaves both `None` -- never a fabricated pair for a row with nothing to add. Both are
    # `None` together, or neither is; a caller never sees one set without the other.
    wait_scans: int | None
    wait_since: str | None


# Comfortably longer than a poll-driven source's own refresh cadence (the *arr poller's default
# is 10s as of 2026-08-21's issue #16, down from 60s -- `core/arrsync.py.ArrSettings.
# poll_interval_s`) -- more than a couple of passes' worth of margin at the new default (this
# constant was not lowered to match; see docs/decisions.md for why leaving it at 150s is the
# conservative call), so one missed refresh never blinks a row out and back. Short enough that a
# row genuinely gone clears within a couple of minutes, not indefinitely. One constant for the
# whole box, not one per source -- the tolerance a user should expect from "briefly not reported"
# is the same regardless of which upstream is doing the reporting.
PREFLIGHT_HOLD_S = 150.0


@dataclass
class _HoldEntry:
    row: PreflightRow
    last_seen: float  # a monotonic clock reading, per source-supplied convention (see below)


class PreflightHold:
    """A flap-tolerant cache of the most recent `PreflightRow` per identity, for one source's own
    scope (`core/arrsync.py` keeps one instance per bound *arr instance id). `update` is the only
    writer, called once per refresh with *every* row that source currently sees; `rows` is the
    only reader. The caller owns what "now" means (`time.monotonic()` throughout this codebase's
    own bounded-retry state, `core/arrsync.py`'s `_InstanceBackoff` and friends) and what counts
    as one identity (a download id, a settle-gate item id, ...) -- this class only manages the
    seen-vs-held-vs-expired bookkeeping once, so every source gets the identical flap tolerance
    without re-deriving it.
    """

    def __init__(self) -> None:
        self._entries: dict[str, _HoldEntry] = {}

    def update(
        self, seen: dict[str, PreflightRow], *, now: float, retired: Iterable[str] = ()
    ) -> None:
        """`seen` is this refresh's complete set, keyed by identity -- every key present gets its
        entry refreshed (row data replaced, `last_seen` bumped to `now`).

        Every key *not* present falls into one of two buckets (2026-08-21, the "a handed-over
        release lingers in Preflight for up to 150s" fix): `retired` is this refresh's own set of
        identities the source knows are gone for a *known* reason -- handed over to a real item,
        for `core/arrsync.py`'s own caller -- and those are deleted immediately, with no smoothing
        at all: there is nothing transient here to hold across. Anything else missing from `seen`
        and absent from `retired` is **merely absent** -- the source has no idea why, which is
        precisely the SABnzbd blank-queue blip this cache exists to absorb -- and keeps today's
        behaviour: held until it has been missing longer than `PREFLIGHT_HOLD_S`, then deleted.
        There is no third state, and a key never needs to appear in both `seen` and `retired` in
        the same call -- a row a source still sees this pass was not, by definition, just retired.
        """
        for key, row in seen.items():
            self._entries[key] = _HoldEntry(row=row, last_seen=now)
        retired_set = set(retired)
        for key in [k for k in self._entries if k not in seen]:
            if key in retired_set or now - self._entries[key].last_seen > PREFLIGHT_HOLD_S:
                del self._entries[key]

    def rows(self) -> list[PreflightRow]:
        return [entry.row for entry in self._entries.values()]

    def items(self) -> list[tuple[str, PreflightRow]]:
        """`rows()` widened with each row's own identity key (2026-08-21, "eviction latency"
        fix) -- for a caller that needs to re-test a held row against something keyed by that
        same identity (`core/arrsync.py.ArrSyncScheduler.preflight_rows`'s own request-time
        retirement re-check, against its last-seen `QueueRecord` per identity) without this
        class exposing its private `_entries` dict directly. Still says nothing about what an
        identity *means* -- that stays the caller's own business, per this module's own
        source-agnostic contract.
        """
        return [(key, entry.row) for key, entry in self._entries.items()]
